cpp/src/strings/copying/concatenate.cu

```diff
@@ -14,140 +14,101 @@
-                bytes = d_offsets[d_view.size()] - d_offsets[0];
-            }
-            return offset_pair{d_view.size(),bytes};
-        },
-        [] __device__ (offset_pair const& lhs, offset_pair const& rhs) {
-            return offset_pair{ lhs.first + rhs.first, lhs.second + rhs.second};
-        });
-    offset_pair last = columns_offsets.back();
-    CUDF_EXPECTS( last.first < std::numeric_limits<size_type>::max(),
-        "total number of strings is too large for a cudf column" );
-    size_type strings_count = static_cast<size_type>(last.first);
-    if( strings_count == 0 )
-        return make_empty_strings_column(mr,stream);
-
-    CUDF_EXPECTS( last.second < std::numeric_limits<size_type>::max(),
-        "total size of strings is too large for a cudf column" );
-    size_type total_bytes = static_cast<size_type>(last.second);
-
-    // set the first set of offsets to 0
-    CUDA_TRY(cudaMemsetAsync( columns_offsets.data().get(), 0, sizeof(offset_pair)));
+            if( d_view.size()==0 )
+                return static_cast<size_t>(0);
+            auto d_offsets = d_view.child(strings_column_view::offsets_column_index).template data<int32_t>();
+            size_t size = d_offsets[d_view.size()+d_view.offset()] - d_offsets[d_view.offset()];
+            return size;
+        }, static_cast<size_t>(0), thrust::plus<size_t>());
+    CUDF_EXPECTS( total_bytes < std::numeric_limits<size_type>::max(), "total size of strings is too large for cudf column" );
 
     // create chars column
-    auto chars_column = create_chars_child_column( strings_count, null_count, total_bytes, mr, stream );
+    auto chars_column = make_numeric_column( data_type{INT8}, total_bytes, mask_state::UNALLOCATED, stream, mr);
     auto d_new_chars = chars_column->mutable_view().data<char>();
+
     // create offsets column
-    auto offsets_column = make_numeric_column( data_type{INT32}, strings_count + 1,
-                                               mask_state::UNALLOCATED, stream, mr);
-    auto d_new_offsets = offsets_column->mutable_view().data<int32_t>();
+    auto offsets_column = make_numeric_column( data_type{INT32}, strings_count + 1, mask_state::UNALLOCATED, stream, mr);
+    auto offsets_view = offsets_column->mutable_view();
+    auto d_new_offsets = offsets_view.data<int32_t>();
 
     // copy over the data for all the columns
-    thrust::for_each_n( execpol->on(stream), thrust::make_counting_iterator<size_type>(0), strings_columns.size(),
-        concatenate_fn{d_views, columns_offsets.data().get(), d_new_offsets+1, d_new_chars} );
-    CUDA_TRY(cudaMemsetAsync(d_new_offsets,0,sizeof(int32_t)));
+    ++d_new_offsets; // skip the first element which will be set to 0 after the for-loop
+    int32_t offset_adjust = 0; // each section of offsets must be adjusted
+    size_type null_count = 0;  // add up the null counts
+    for( auto column = strings_columns.begin(); column != strings_columns.end(); ++column )
+    {
```

Comment:
Consider a functor that: 

1. encapsulates the body of this big loop; 
2. provides a constructor taking all the necessary arguments;
3. and apply a `std::for_each`, or `std::transform` algorithm on the collection of columns using this functor;

Separating responsibilities like that would make the code more modular, readable, and easier to maintain in the future. Perhaps place such a functor in a `<name>_detail.cuh` header. If, in the future, the functor's functionality must change - slightly or more dramatically - then this code (the "client" code) would not have to change, just the functor. Hence, providing a neat separation of concerns and modularity.

Reply:
I just reverted this code to a previous version in order to fix the issue referenced.
I believe Trevor is to rework all the concatenate code for performance.
I'd hate to refactor all of this only for him to rip it all out with a different implementation.
Is this change a requirement to approve this PR?

Reply:
No. Just a suggestion. Please discard and feel free to resolve conversation.